extends Node2D

## ç©å®¶å¯¹è±¡
@export
var player: Player

## æ¸¸æˆæ˜¯å¦ç»“æŸï¼Œé»˜è®¤ï¼šfalse
var _is_game_over: bool = false

## è§†çª—è®¾å®šå¤§å°
var _viewport_size = Vector2(416.0, 260.0)

@onready
var boss: Boss = $SubViewportContainer/SubViewport/Boss

@onready
var viewport: SubViewport = $SubViewportContainer/SubViewport

@onready
var viewport_container: SubViewportContainer = $SubViewportContainer

@onready
var life_container: HBoxContainer =\
	$SubViewportContainer/SubViewport/HudContainer/HeroProfileContainer/LifeContainer

@onready
var blood_progress: TextureProgressBar =\
	$SubViewportContainer/SubViewport/HudContainer/HeroProfileContainer/HeroBloodProgressBar

@onready
var player_packed_scene: PackedScene = preload('res://sprites/tscns/player.tscn')

func _ready() -> void:
	set_physics_process(true)
	_update_ui_display() #æ›´æ–°UIæ˜¾ç¤º
	get_window().size_changed\
		.connect(_update_ui_display)
	GlobalSignals.on_game_over\
		.connect(_on_game_over)
	GlobalSignals.on_player_dead\
		.connect(_on_player_dead)
	boss.spy_player = player
	boss.on_boss_die.connect(_on_boss_die)

## æ›´æ–°UIæ˜¾ç¤º
func _update_ui_display() -> void:
	var visible_rect = get_viewport().get_visible_rect()
	var visible_size = visible_rect.size
	var scaler = minf(visible_size.x / _viewport_size.x,\
		visible_size.y / _viewport_size.y)
	var target_width = _viewport_size.x * scaler
	var target_height = _viewport_size.y * scaler
	var offset_x = (visible_size.x - target_width) /2.0
	var offset_y = (visible_size.y - target_height) / 2.0
	viewport_container.scale = Vector2(scaler, scaler)
	viewport_container.global_position = Vector2(offset_x, offset_y)

func _physics_process(_delta: float) -> void:
	if player: #å¦‚æœç©å®¶å¯ç”¨
		#region è®¾ç½®ç©å®¶å¯è¡Œèµ°çš„èŒƒå›´
		player.walk_area_x = viewport_container.size.x
		player.walk_area_y = viewport_container.size.y
		#endregion
		#region æ›´æ–°è¡€æ¡ä¿¡æ¯
		blood_progress.value =\
			(player.life_blood / player.life_blood_max) * 100.0
		var life_count = life_container.get_child_count()
		var diff_count = GlobalConfigs.player_life_count - life_count
		if diff_count > 1:
			var texture = TextureRect.new()
			texture.texture =\
				load('res://assets/ui/ui_medal_life.png') as Texture2D
			life_container.add_child(texture)
		elif life_count > 0:
			var last_child = life_container.get_child(0)
			life_container.remove_child(last_child)
		#endregion

## åˆ›å»ºç©å®¶
func _create_player(location: Vector2) -> Player:
	var _player = player_packed_scene.instantiate() as Player
	_player.global_position = location
	return _player

## bossè¢«ç©å®¶æ¶ˆç­
func _on_boss_die() -> void:
	print('Bossè¢«æ¶ˆç­ï¼ä½ èµ¢äº†ï¼')

## ç©å®¶è¢«æ¶ˆç­
func _on_player_dead(location: Vector2) -> void:
	var life_count =\
		GlobalConfigs.player_life_count
	if life_count >= 1:
		life_count = life_count - 1
	if life_count > 0:
		if player: 
			player.free()
			player = null
		player = _create_player(\
			Vector2(location.x, location.y - 60.0))
		viewport.add_child(player)
		if boss and not boss._is_die: #å¦‚æœbossè¿˜å­˜åœ¨
			get_tree().create_timer(0.5)\
				.timeout.connect(func(): boss.spy_player = player)
	GlobalConfigs.player_life_count = life_count
	if not _is_game_over: _on_game_over() #æ¸¸æˆç»“æŸ

## æ¸¸æˆç»“æŸäº†
func _on_game_over() -> void:
	if not _is_game_over: return
	_is_game_over = true #æ ‡è®°æ¸¸æˆå·²ç»ç»“æŸ
	print('æ¸¸æˆç»“æŸå•¦ï¼ğŸ˜Š')
